File: E3Box/src/com/cubic/e3box/sip/SipService.java
package com.cubic.e3box.sip;

import java.util.Timer;
import java.util.TimerTask;
import java.util.UUID;
import org.json.JSONException;
import org.json.JSONObject;
import com.cubic.e3box.CarrierApp;
import com.cubic.e3box.Constants;
import com.cubic.e3box.ICallCenter;
import com.cubic.e3box.MainReceive;
import com.cubic.e3box.MainReceive.netEventHandler;
import com.cubic.e3box.sip.ThirNativeServiceDef;
import com.cubic.e3box.utils.CommonUtils;
import com.cubic.e3box.utils.HttpsRequestUtils;
import android.os.IBinder;
import android.os.RemoteException;
import android.os.ServiceManager;
import android.os.SystemProperties;
import android.util.Base64;
import android.util.Log;

public class SipService implements netEventHandler {
	public static final String TAG="SipService";
	
	private boolean mIsSipRegistered = false;
	private int mNetType = NET_TYPE_NONE;
	
	private static final int NET_TYPE_NONE = 0;
	private static final int NET_TYPE_WIFI = 1;
	private static final int NET_TYPE_WIRED =2;
	
	// command to service
	private static final int SIP_CMD_ECHO = 1;
	private static final int SIP_CMD_DAIL = 2;
	private static final int SIP_CMD_HANGUP =3;
	private static final int SIP_CMD_DTMF = 4;
	private static final int SIP_CMD_ACCEPT = 5;
	private static final int SIP_CMD_GET_STAT = 6;
	private static final int SIP_CMD_REGISTER = 7;
	
	// command from service
	private static final int SIP_RPT_STAT = 100;
	private static final int SIP_RPT_INCOMMING_CALL = 101; 
	private static final int SIP_RPT_HANGUP = 102;
	private static final int SIP_RPT_REGISTER_STAT = 103;
	private static final int SIP_RPT_ACCEPT = 104;
	
	public SipService(){
		MainReceive.mNetworkListeners.add(this);
	}
	
	private static void logout(String log) {
		Log.e(TAG, TAG + " " + log );
	}
	
	private ThirNativeServiceDef.BpThirNativeService mNativeService = null;
	private ThirNativeServiceDef.BnThirNativeService mNativeServiceCallback = new ThirNativeServiceDef.BnThirNativeService() {

		// data from sipd
		@Override
		public int sendData(byte[] data) {
			if( mCallCenter != null )
			{
				mCallCenter.onSoundData(data);
			}
			return 0;
		}

		// report from sipd
		@Override
		public int sendCmd(int code, String info) {
			switch(code)
			{
			case SIP_RPT_STAT:
				// do nothing
				break;

			case SIP_RPT_INCOMMING_CALL:
				if( mCallCenter != null )
				{
					String[] strInfoTmp = info.split("::");
					if(strInfoTmp.length == 2){
						String strNumber = strInfoTmp[0];
						String strXwsse = strInfoTmp[1];
						
						if(CommonUtils.checkPasswordDigest(strXwsse,CommonUtils.PASSWORD_TYPE_PSTN)){
							mCallCenter.onIncommingCall(strNumber);
						}else{
							Log.e(TAG,"checkPasswordDigest error");
						}
					}else{
						mCallCenter.onIncommingCall(info);
					}
				}
				break;

			case SIP_RPT_HANGUP:
				if( mCallCenter != null )
				{
					mCallCenter.onHangup();
				}
				break;
				
			case SIP_RPT_REGISTER_STAT:
				logout("SIP_RPT_REGISTER_STAT regCode = " + info);
				mIsSipRegistered = info.equals("200");
				break;
			
			case SIP_RPT_ACCEPT:
				if( mCallCenter != null )
				{
					mCallCenter.onAccept();
				}
				break;
			};
			return 0;
		}

		@Override
		public int regBinder(IBinder binder) throws RemoteException {
			// empty function, do nothing
			return 0;
		}
	};
	
	private TimerTask mTimerTask = new TimerTask() {
		@Override
		public void run() {
			// if not connected
			if(mNativeService == null){
				logout("ServiceManager.TimerTask.run: try to connect service ..");
				// init service proxy
				IBinder binder = ServiceManager.getService(ThirNativeServiceDef.DESCRIPTOR);
				if( binder == null ) {
					logout("ServiceManager.getService failed !");
					return;
				}

				mNativeService = new ThirNativeServiceDef.BpThirNativeService(binder);
				if( mNativeService == null ) {
					logout("mNativeService create Failed!");
					return;
				}

				// regiester service
				try
				{
					int ret = mNativeService.regBinder(mNativeServiceCallback);
					logout("mNativeService.doAction.regBinder ret="+ret);
				}
				catch( Exception e )
				{
					e.printStackTrace();
				}
			} else {
				// test if service still online
				// if service offline, reconnect it
				if( sendCmd(SIP_CMD_ECHO) < 0 ) {
					logout("ServiceManager.TimerTask.run: lost service !");
					mNativeService = null;
				}
			}
		}
	};
	
	private Timer mTimer = new Timer();

	private ICallCenter mCallCenter = null;
	
	private int sendCmd( int code, String info ) {
		int ret = 0;

		if(mNativeService == null) {
			return -100;
		}
		
		try
		{
			ret = mNativeService.sendCmd(code ,info);
		}
		catch( Exception e )
		{
			e.printStackTrace();
			ret = -99;
		}
		return ret;
	};
	private int sendCmd( int code ) {
		return sendCmd(code, "");
	};

	public boolean init(ICallCenter callCenter) {
		logout("init");
		mTimer.schedule(mTimerTask, 100, 2000);
		mCallCenter = callCenter;
		return true;
	};

	public void deinit() {
		mTimer.cancel();
	};

	public boolean dial(String number) {
		if( 0 > sendCmd(SIP_CMD_DAIL, number) )
		{
			return false;
		}
		return true;
	};

	public boolean dtmf(String number) {
		if( 0 > sendCmd(SIP_CMD_DTMF, number) )
		{
			return false;
		}
		return true;
	};

	public void hangup() {
		sendCmd(SIP_CMD_HANGUP);
	};

	public void accept() {
		sendCmd(SIP_CMD_ACCEPT);
	};

	public void register(String regInfo){
		sendCmd(SIP_CMD_REGISTER,regInfo);
	}
	
	public void writeSound( byte[] data ) {
		if( mNativeService == null ) {
			return;
		}

		try {
			mNativeService.sendData(data);
		} catch (RemoteException e) {
			e.printStackTrace();
		}
	}
	
	@Override
	public void onNetChange() {
    	logout("wifi state = "+CarrierApp.getNetWifiState()+" ; " +
    			"ethernet state = "+CarrierApp.getNetEthernetState());
    	if(CarrierApp.getNetWifiState() == true || CarrierApp.getNetEthernetState() == true){
			String domain = SystemProperties.get(Constants.BLE_PROP_sip_domain, Constants.BLE_DEFAULT_VALUE_NULL );
			String acc_user = SystemProperties.get(Constants.BLE_PROP_sip_acc_user, Constants.BLE_DEFAULT_VALUE_NULL );
			String acc_pass = SystemProperties.get(Constants.BLE_PROP_sip_acc_pass, Constants.BLE_DEFAULT_VALUE_NULL );
			if(domain.equals(Constants.BLE_DEFAULT_VALUE_NULL) || 
					acc_user.equals(Constants.BLE_DEFAULT_VALUE_NULL) || 
					acc_pass.equals(Constants.BLE_DEFAULT_VALUE_NULL)){
				logout("Error:register infomation is not usefull");
				return;
			}
			String id = "sip:"+acc_user+"@"+domain+";transport=tcp";
			String registrar = "sip:"+domain+";transport=tcp";
			String proxy = "sip:sip.test.e3phone.info:38060;transport=tcp";
			logout("sip register : "+id+"&"+registrar+"&"+"*"+"&"+acc_user+"&"+acc_pass+"&"+proxy);
			register(id+"&"+registrar+"&"+"*"+"&"+acc_user+"&"+acc_pass+"&"+proxy);
    		if(CarrierApp.getNetWifiState() == true){
    			mNetType = NET_TYPE_WIFI;
    		}else{
    			mNetType = NET_TYPE_WIRED;
    		}
    		/***This is a example for Https Request***
			JSONObject jsonReq = new JSONObject();  
			try {
				jsonReq.put("X-Push-Call-ID", UUID.randomUUID().toString());
				jsonReq.put("From", "075-123-4567");
				jsonReq.put("To", "03-123-4567");
			} catch (JSONException e) {
				e.printStackTrace();
			}
			HttpsRequestUtils httpsReq = new HttpsRequestUtils(HttpsRequestUtils.HTTPS_REQ_CALL_All_GROUP_MEMBERS, jsonReq);
			httpsReq.httpsPostData();*/
    	}else{
    		mNetType = NET_TYPE_NONE;
    	}
	};
}

File: pjproject-2.4/pjsipd/SipService.cc
#ifndef _SIP_SERVICE_CC_
#define _SIP_SERVICE_CC_ 1

#include <utils/RefBase.h>
#include <binder/IInterface.h>
#include <binder/Parcel.h>
#include <binder/IPCThreadState.h>
#include <binder/ProcessState.h>
#include <binder/IServiceManager.h>
#include <cutils/properties.h>
#include <CSafeQueue.cc>
#include "ThirAndroidLog.h"
#include "IThirNativeService.cc"
#include "SipCaller.h"

ISipUser* SipCaller::m_user = NULL;

class SipService : public BnThirNativeService, public ISipUser
{
public:
	enum
	{
		// command from user
		SIP_CMD_ECHO = 1,
		SIP_CMD_DAIL,
		SIP_CMD_HANGUP,
		SIP_CMD_DTMF,
		SIP_CMD_ACCEPT,
		SIP_CMD_GET_STAT,
		SIP_CMD_REGISTER,
		
		// command to user
		SIP_RPT_STAT = 100,
		SIP_RPT_INCOMMING_CALL,
		SIP_RPT_HANGUP,
		SIP_RPT_REGISTER_STAT,
		SIP_RPT_ACCEPT,
	};

private:
	BpThirNativeService* m_user;
	SipCaller m_caller;

	void strToC16( const char* str, char16_t* str16, int len )
	{
		const char* p = str;
		char16_t*   q = str16;
		for(int i = 0; i < len && *p; i++, p++, q++ )
		{
			*q = *p;
		}
	};

	void strToC( const char16_t* str16, char* str, int len )
	{
		const char16_t* p = str16;
		char*           q = str;
		for(int i = 0; i < len && *p; i++, p++, q++ )
		{
			*q = (char)((*p)&0xFF);
		}
	};
	
public:
	
	SipService(int argc, char *argv[])
		: m_user(NULL)
		, m_caller(argc,argv,this)
	{};
	
	~SipService()
	{
		if(m_user)
		{
			delete m_user;
		}
	};

	// **************************** call by user ********************************

	// command from user
	virtual int32_t sendCmd(int32_t code, const char16_t* info)
	{
		static const int INFO_LEN_MAX = 1024;
		char s_info[INFO_LEN_MAX+4] = {0};

		
		switch( code )
		{
		case SIP_CMD_ECHO:
			return 0;

		case SIP_CMD_DAIL:
			strToC(info, s_info, INFO_LEN_MAX);
			m_caller.dail(s_info);
			LOGD("SipService::sipd get cmd of SIP_CMD_DAIL , dailNum = %s",s_info);
			return 0;

		case SIP_CMD_HANGUP:
			LOGD("SipService::sipd get cmd of SIP_CMD_HANGUP");
			m_caller.hangup();
			return 0;

		case SIP_CMD_DTMF:
			LOGD("SipService::sipd get cmd of SIP_CMD_DTMF");
			return 0;

		case SIP_CMD_ACCEPT:
			LOGD("SipService::sipd get cmd of SIP_CMD_ACCEPT");
			m_caller.accept();
			return 0;

		case SIP_CMD_GET_STAT:
			LOGD("SipService::sipd get cmd of SIP_CMD_GET_STAT");
			return m_caller.getStat();

		case SIP_CMD_REGISTER:
			strToC(info, s_info, INFO_LEN_MAX);
			LOGD("SipService::sipd get cmd of SIP_CMD_REGISTER , regInfo = %s", s_info);
			m_caller.sipRegister(s_info);
			return 0;
			
		default:
			break;
		};
		return -2;
	};

	// data from user
	virtual int32_t sendData(int32_t size, const void* data)
	{
		m_caller.writeSound(data, size);
		return 0;
	};

	virtual int32_t regBinder(sp<IBinder> &binder)
	{
		LOGD("SipService::regBinder()");
		if( m_user == NULL )
		{
			m_user = new BpThirNativeService(binder);
		}
		return 0;
	};
	
	// ******************** call by sip caller *********************
	virtual void onRegisterStat(int regCode)
	{
		static const int REG_CODE_LEN_MAX = 64;
		char strRegCode[REG_CODE_LEN_MAX];
		sprintf(strRegCode,"%d",regCode);
		char16_t s_strRegCode[REG_CODE_LEN_MAX+4] = {0};
		if(m_user)
		{
			strToC16(strRegCode, s_strRegCode, REG_CODE_LEN_MAX);
			m_user->sendCmd(SIP_RPT_REGISTER_STAT, s_strRegCode);
		}
	}
	
	virtual void onIncommingCall( const char *number )
	{
		static const int NUMBER_LEN_MAX = 2048;
		char16_t s_number[NUMBER_LEN_MAX+4] = {0};
		
		if(m_user)
		{
			strToC16(number, s_number, NUMBER_LEN_MAX);
			m_user->sendCmd(SIP_RPT_INCOMMING_CALL, s_number);
		}
	};

	virtual void onHangup()
	{
		char16_t empty_str[] = {'\0','\0'};

		if(m_user)
		{
			m_user->sendCmd(SIP_RPT_HANGUP, empty_str);
		}
	};

	virtual void onAccept()
	{
		char16_t empty_str[] = {'\0','\0'};

		if(m_user)
		{
			m_user->sendCmd(SIP_RPT_ACCEPT, empty_str);
		}
	};
	
    virtual void onSound(const void* data, int size)
    {
    	if(m_user)
		{
			m_user->sendData(size, data);
		}
    };
};
#endif
